In [2]:
!pip3 install tiktoken

  Using cached requests-2.34.2-py3-none-any.whl.metadata (4.8 kB)
  Using cached charset_normalizer-3.4.7-cp313-cp313-win_amd64.whl.metadata (41 kB)
  Using cached urllib3-2.7.0-py3-none-any.whl.metadata (6.9 kB)
   ---------------------------------------- 0.0/874.8 kB ? eta -:--:--
   ----------------------------------- ---- 786.4/874.8 kB 4.3 MB/s eta 0:00:01
   ---------------------------------------- 874.8/874.8 kB 3.3 MB/s  0:00:00
Using cached requests-2.34.2-py3-none-any.whl (73 kB)
Using cached charset_normalizer-3.4.7-cp313-cp313-win_amd64.whl (158 kB)
Using cached urllib3-2.7.0-py3-none-any.whl (131 kB)

   ---------------------------------------- 0/7 [urllib3]
   ---------------------------------------- 0/7 [urllib3]
   ---------------------------------------- 0/7 [urllib3]
   ---------------------------------------- 0/7 [urllib3]
   ---------------------------------------- 0/7 [urllib3]
   ---------------------------------------- 0/7 [urllib3]
   ---------------------------

## BYTE PAIR ENCODING

In [4]:
import importlib
import tiktoken

print("tiktoken version:", importlib.metadata.version("tiktoken"))  

tiktoken version: 0.13.0


In [5]:
tokenizer = tiktoken.get_encoding("gpt2")

In [19]:
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)

integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print("Encoded integers:", integers)
print(type(integers))

for integer in integers:
    print(integer, tokenizer.decode([integer]), sep = " -> ")


Encoded integers: [15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]
<class 'list'>
15496 -> Hello
11 -> ,
466 ->  do
345 ->  you
588 ->  like
8887 ->  tea
30 -> ?
220 ->  
50256 -> <|endoftext|>
554 ->  In
262 ->  the
4252 ->  sun
18250 -> lit
8812 ->  terr
2114 -> aces
1659 -> of
617 ->  some
34680 -> unknown
27271 -> Place
13 -> .


We can make two noteworthy observations based on the token IDs and decoded text above.

First, the <|endoftext|> token is assigned a relatively large token ID, namely, 50256.

In fact, the BPE tokenizer, which was used to train models such as GPT-2, GPT-3, and the original model used in ChatGPT, has a total vocabulary size of 50,257, with <|endoftext|> being assigned the largest token ID.

Second, the BPE tokenizer above encodes and decodes unknown words, such as "someunknownPlace" correctly.

The BPE tokenizer can handle any unknown word. How does it achieve this without using <|unk|> tokens?

The algorithm underlying BPE breaks down words that aren't in its predefined vocabulary into smaller subword units or even individual characters.

The enables it to handle out-ofvocabulary words.

So, thanks to the BPE algorithm, if the tokenizer encounters an unfamiliar word during tokenization, it can represent it as a sequence of subword tokens or characters

Let us take another simple example to illustrate how the BPE tokenizer deals with unknown tokens

In [20]:
strings = tokenizer.decode(integers)

print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


In [22]:
integers = tokenizer.encode("Akwirw ier")
print(integers)

for integer in integers:
    print(integer, tokenizer.decode([integer]), sep = " -> ")

[33901, 86, 343, 86, 220, 959]
33901 -> Ak
86 -> w
343 -> ir
86 -> w
220 ->  
959 -> ier


## Creating Input target pair

In [28]:
with open("data/verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print('Sample encoded tokens:', enc_text[:10])
print('Total encoded tokens:', len(enc_text))

Sample encoded tokens: [40, 550, 1464, 1807, 3619, 402, 271, 10899, 2138, 257]
Total encoded tokens: 5373
